internal/command/machine/select.go

```diff
@@ -0,0 +1,121 @@
+package machine
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strings"
+
+	"github.com/superfly/flyctl/api"
+	"github.com/superfly/flyctl/client"
+	"github.com/superfly/flyctl/flaps"
+	"github.com/superfly/flyctl/internal/appconfig"
+)
+
+func selectOneMachine(ctx context.Context, machineID string) (*api.Machine, context.Context, error) {
+	return selectOneMachineImpl(ctx, machineID, nil)
+}
+
+func selectOneMachineFromApp(ctx context.Context, machineID string, app *api.AppCompact) (*api.Machine, context.Context, error) {
+	if app == nil {
+		return nil, nil, errors.New("tried to select a machine from a nil app; this is a bug")
+	}
+	return selectOneMachineImpl(ctx, machineID, app)
+}
+
+func selectOneMachineImpl(ctx context.Context, machineID string, app *api.AppCompact) (*api.Machine, context.Context, error) {
+	var err error
+
+	if app != nil {
+		ctx, err = buildContextFromApp(ctx, app)
+	} else {
+		ctx, err = buildContextFromAppNameOrMachineID(ctx, machineID)
+	}
```

Comment:
I'm not sure I see the need for three functions here. Since `selectOneMachineImpl` already handles whether app is nil or not, it could just be one function, or if you want the enforcement that app is not nil in some cases, two functions.